gadget/mountedfilesystem.go

```diff
@@ -179,10 +198,555 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+		}
+	}
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) backupOrCheckpointFile(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+	srcName, dstName, backupBase := f.entryPaths(dstRoot, source, target, backupDir)
+
+	backupName := backupBase + ".backup"
+	sameStamp := backupBase + ".same"
+	preserveStamp := backupBase + ".preserve"
+
+	if !osutil.FileExists(dstName) {
+		// destination does not exist and will be created when writing
+		// the udpate, no need for backup
+		return nil
+	}
+
+	if osutil.FileExists(backupName) || osutil.FileExists(sameStamp) {
+		// file already checked, either has a backup or is the same as
+		// the update, move on
+		return nil
+	}
+
+	if strutil.SortedListContains(preserveInDst, dstName) {
+		// file is to be preserved, create a relevant stamp
+
+		if !osutil.FileExists(dstName) {
+			// preserve, but does not exist, will be written anyway
+			return nil
+		}
+		if osutil.FileExists(preserveStamp) {
+			// already stamped
+			return nil
+		}
+		// make a stamp
+		if err := makeStamp(preserveStamp); err != nil {
+			return fmt.Errorf("cannot create preserve stamp: %v", err)
+		}
+		return nil
+	}
+
+	// try to find out whether the update and the existing file are
+	// identical
+
+	orig, err := os.Open(dstName)
+	if err != nil {
+		return fmt.Errorf("cannot open destination file: %v", err)
+	}
+
+	// backup of the original content
+	backup, err := newStampFile(backupName)
+	if err != nil {
+		return fmt.Errorf("cannot create backup file: %v", err)
+	}
+	// becomes a backup copy or a noop if canceled
+	defer backup.Commit()
+
+	// checksum the original data while it's being copied
+	origHash := crypto.SHA1.New()
+	htr := io.TeeReader(orig, origHash)
+
+	_, err = io.Copy(backup, htr)
+	if err != nil {
+		backup.Cancel()
+		return fmt.Errorf("cannot backup original file: %v", err)
+	}
+
+	// digest of the update
+	updateDigest, _, err := osutil.FileDigest(srcName, crypto.SHA1)
+	if err != nil {
+		backup.Cancel()
+		return fmt.Errorf("cannot checksum update file: %v", err)
+	}
+	// digest of the currently present data
+	origDigest := origHash.Sum(nil)
+
+	if bytes.Equal(origDigest, updateDigest) {
+		// mark that files are identical and update can be skipped, no
+		// backup is needed
+		if err := makeStamp(sameStamp); err != nil {
+			return fmt.Errorf("cannot create a checkpoint file: %v", err)
+		}
+
+		// makes the previous commit a noop
+		backup.Cancel()
+		return nil
+	}
+
+	// update will overwrite existing file, a backup copy is created on
+	// Commit()
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) backupVolumeContent(volumeRoot string, content *VolumeContent, preserveInDst []string, backupDir string) error {
+	if err := checkContent(content); err != nil {
+		return err
+	}
+
+	srcName := f.entrySourcePath(content.Source)
+
+	if err := f.checkpointPrefix(volumeRoot, content.Target, backupDir); err != nil {
```

Comment:
I'm still not sure I understand why the call patterns for checkpointPrefix and rollbackPrefix don't match

Reply:
This is cleaned up now. For completeness I introduced a `makePrefix` call for the update path as well.

While doing so I caught some edge cases when writing files to a nested prefix that does not exist or empty directories to non-existent ones. The tests were updated to reflect this.